### ISSUE TITLE
Changes for Descriptions

### ISSUE THREAD

**Issue by terracoda** [phetsims/balloons-and-static-electricity#148]
Version tested: http://www.colorado.edu/physics/phet/dev/html/balloons-and-static-electricity/1.2.0-reader.1/balloons-and-static-electricity_en.html?reader

This version of the Reader did not work on Mac Book Pro with Yosemite - voiced verbalizations were extremely intermittent. I have been having trouble with my Mac Book Pro (it is old (early 2011)), so it could be the computer. It’s been crashing lately.

On the Mac Air with the Reader mode worked. I have made two tests (working on third). In general it seems to work. I listened for correct descriptions and trying to understand how new interaction affects the descriptions. I also did some comparisons with VoiceOver. There are differences, but they are not recorded here. 

# Test One: Went through sim with down cursor key
                          
### List item 3 describing charge information is missing balloon charge info.
- [x] Should be: "Sweater and wall have many pairs of negative and positive charges, the balloon has just a few pairs. Charges can change as you play and learn about static electricity."

### There was no navigation cue. Navigation cue should be:
- [x] "Tab to Play with balloon." Maybe it was left off on purpose? See comments in issue #135? I think we may need it, though.

### Remove Wall, button
- [x] On-demand help should not include the words ”Toggle to" and has been simplified to:
"Conduct experiments with or without the Wall."
- [x] The help text is currently read out automatically. Please make sure that the `aria-describedby` is un-associated. Users will get help text "on-demand" with the down cursor key. 

### Two-balloon experiment, button
- [x] On-demand help text is missing. The new simplified help-text is: 
"Play with two balloons, or just one."

Note that the word "toggle" may be verbalized naturally by some screen readers due to the type of control. Also note, ARIA 1.1 has a new switch role. I am not sure if screen readers recognize that role yet, but we could try changing the role on the interaction to ”switch" at some point. The new role is like a checkbox, but should verbalize “On” and “Off” rather than checked and unchecked.

### Reset Balloon(s), button
- [x] On-demand help text should not include the word "Select" and has been changed to: 
- "Return balloon to start position and an uncharged state." 
- "Return balloons to start positions and an uncharged state."

# Test Two: I did a Tab through
## Grab Yellow Balloon, button
- [x] There is no on-demand help text, please add the following:
“Get ready to drag balloon. Once grabbed press W, A, S, or D key to drag up, left, down, or right.”

### When balloon is grabbed, I hear the following:
“Yellow Balloon, draggable, undefined Use W, A, S, D keys to grab and drag balloon up, left, down, and right. Tab for next object. Press H for keyboard commands and help.” 
- [x] Question: Do we know where the word "undefined" is coming from? What exactly is undefined? Maybe we can define it.
- [x] Please remove the cue to “Tab for next object.” 
- [x] The description of the grabbed balloon can be changed to a list consisting of:
- balloon’s current position upon grabbing
- balloon charge info
- description of nearest object (sweater or wall)
-  then finally the navigation cues for help and release. 

- [x] Please use the following shortened navigation cue: 
"Press W, A, S, or D key to drag balloon. Spacebar to release it. H key for more balloon hot keys."

Note I removed the lengthy “up, left, down, or right” because we now have instant feedback, so let’s see if the shortened cue can get users dragging.

### Released Balloon 
- [x] There is currently no indication that the balloon is released. We need to add an interaction alert that first announces the release, and then perspective 3 description. All this before the “Grab Yellow Balloon, button” is read out. Hopefully, this is possible.

Example of the release description list:
- State and position: Negatively charged Balloon released from [center of Play Area].
- Movement: Attracted [quickly] back to sweater.
* Resting position: Now, sticks to [upper right side] of sweater.
- Description of nearest object: Sweater has a net positive charge, many more positive charges than negative ones.
- Balloon has a net negative charge, many more negative charges than positive ones.
- Finally, “Grab balloon button”.

# Now working on Test Three: Dragging descriptions.

**Comment by jessegreenberg** [comment]
This is a great review @terracoda, thanks for doing this.  I will begin working on this.

**Comment by jessegreenberg** [comment]
Commit above adds

> Should be: "Sweater and wall have many pairs of negative and positive charges, the balloon has just a few pairs. Charges can change as you play and learn about static electricity."

However, this content is static for now.

**Comment by jessegreenberg** [comment]
Commit above fixes

> "Tab to Play with balloon." Maybe it was left off on purpose? See comments in issue #135? I think we may need it, though.

Not left out on purpose, bug in code prevented this content from being added to PDOM.

**Comment by jessegreenberg** [comment]
Commits above implement changes to the 'Add or Remove Wall' button.

**Comment by jessegreenberg** [comment]
Commit above fixes

> On-demand help text is missing. The new simplified help-text is: "Play with two balloons, or just one."

Thanks for the reminder about the 'toggle'.  This should be supported by the prototype screen reader.

> Note that the word "toggle" may be verbalized naturally by some screen readers due to the type of control.

**Comment by terracoda** [comment]
@jessegreenberg, regarding screen reader verbalizations, they vary and we don't have to copy vague descriptions :-). If we have control, it would be better to say,  "Two-balloon experiment, switch, off" which patterns "Two-balloon experiment, toggle button, not pressed". The language for a switch control is much clearer.

**Comment by jessegreenberg** [comment]
@terracoda, great!  We can use "switch, off" instead.  Is there different ARIA markup that should be used?  Or is that the desired output from the screen reader for 'aria-pressed'?

**Comment by terracoda** [comment]
@jessegreenberg, the code might be more like a checkbox, but I am not sure. What's the HTML for a toggle button?

**Comment by jessegreenberg** [comment]
A toggle button is just a button element with the 'aria-pressed' attribute like
```html
<button id="abswitch-17-61-513-564-557-544-539-543" aria-pressed="true">‪Two Balloon Experiment‬</button>
```
With 'aria-pressed' the screen reader may read 'Toggle' before 'Button' or 'Pressed' if 'aria-pressed'' == ```true```.

**Comment by terracoda** [comment]
`<input id="abswitch-17-61-513-564-557-544-539-543" type="checkbox" aria-checked>`

**Comment by terracoda** [comment]
And we need a name or label, of course.

**Comment by jessegreenberg** [comment]
What is the difference between 
```
<input id="abswitch-17-61-513-564-557-544-539-543" type="checkbox" aria-checked='false'>
```
and a normal checkbox?  Is there another role that should be used to give it 'switch' behavior?

**Comment by terracoda** [comment]
Added 
```html role="switch"
```
But we have to keep in mind that standard screen readers might not recognize this role yet. It's brand new.

**Comment by jessegreenberg** [comment]
https://github.com/phetsims/balloons-and-static-electricity/issues/148#issuecomment-221141956 <--- perfect thanks!

**Comment by terracoda** [comment]
I forgot the role, the essential ingredient, duh!

**Comment by terracoda** [comment]
@jessegreenberg, have you noticed that if you have the balloon grabbed and you accidentally press the arrow keys, you can no longer interact with the balloon? If I press the arrow keys while balloon is grabbed, I have to tab away and then grab it again to keep interacting. This will be a problem for screen reader users. If the screen reader goes silent, it becomes a little disorientating.

If I get to work early enough in the morning, I'm going to test with JAWS and NVDA. I installed them on an extra Windows 7 PC at work.

**Comment by jessegreenberg** [comment]
> @jessegreenberg, have you noticed that if you have the balloon grabbed and you accidentally press the arrow keys, you can no longer interact with the balloon?

No, I haven't, but thanks for bringing that up. That will be an issue for sure.

**Comment by jessegreenberg** [comment]
> When balloon is grabbed, I hear the following:
“Yellow Balloon, draggable, undefined

It looks like description is missing 'grabbed'.  Perhaps safari doesn't support 'aria-grabbed'?  I will look into it.

**Comment by jessegreenberg** [comment]
> The description of the grabbed balloon can be changed to a list consisting of:
 - balloon’s current position upon grabbing
 - balloon charge info
 - description of nearest object (sweater or wall)
 - then finally the navigation cues for help and release.

@terracoda since the balloon is an 'appliciation', the user won't be able to find list items with the arrow keys.  How will this content be found?

**Comment by jessegreenberg** [comment]
Adding a checkbox for the 'release balloon' alert.

**Comment by jessegreenberg** [comment]
@terracoda, given your comment 
> But we have to keep in mind that standard screen readers might not recognize this role yet. It's brand new.

Would you like to proceed with a 'switch' or keep the toggle button?  The prototype reader can support ```role=switch```.

**Comment by terracoda** [comment]
@jessegreenberg I think it is worth testing. The specification is in working draft. I have no idea how fast new standards are implemented. I'd like to test it to see what the control could sound like.

**Comment by terracoda** [comment]
@jessegreenberg, regarding the a list and aria-describedby, we might be able to use multiple ids to associate multiple descriptions. I found the following resource, but I am still not sure how it will work [Using ARIA](https://www.w3.org/TR/WCAG20-TECHS/ARIA1.html)

**Comment by jessegreenberg** [comment]
@terracoda, sounds good.  I will try it a switch.

**Comment by jessegreenberg** [comment]
> There is currently no indication that the balloon is released. We need to add an interaction alert that first announces the release, and then perspective 3 description. All this before the “Grab Yellow Balloon, button” is read out. Hopefully, this is possible.

I am concerned that many screen readers will treat the shift of focus as an 'assertive' update, and 'Grab Yellow Balloon' button will be read over all other 'aria-live' messages when the simulation shifts focus to the Balloon Button.

Perhaps we can think of another solution that produces the same interaction for the user, but does not involve two HTML elements.

**Comment by jessegreenberg** [comment]
I am not sure if screen readers support this, but we could keep the same HTML element by using ARIA roles.  Then the role would switch between 'button' and 'application' depending on whether or not the balloon is being dragged.  Screen reader support for button isn't great, but perhaps this would let us have control over the order of updates since we are not manipulating focus.

**Comment by jessegreenberg** [comment]
https://github.com/phetsims/balloons-and-static-electricity/issues/148#issuecomment-221402249 I am going to give this a shot.

**Comment by terracoda** [comment]
@jessegreenberg 1.2.0-reader.2 is the latest, right?

I'm experiencing issues with NVDA and FF, and NVDA and Chrome on a Windows 7 machine, Intel Core Duo CPU 2.66GHz, 64 but with 8.00 GB of RAM.

This system should be good enough, no? 
I've updated to the latest Firefox and the latest Chrome.

**Comment by jessegreenberg** [comment]
@terracoda, yes, I wouldn't expect any issues.  Does NVDA run well outside of the simulation?

**Comment by terracoda** [comment]
@jessegreenberg, I was using NVDA to do a website assessment and it seemed to work fine. I'll try again tomorrow. When I shutdown the machine, Windows started installing several updates.

**Comment by terracoda** [comment]
@jessegreenberg, Re our discussion about lists. It's possible, even likely, that the list structure I would like for the balloon description won't work. However, according to the specification we can have multiple aria-describedby id's associates with an element. Given this information, we should be able to have multiple sentences, even if they are not in a list structure.

**Comment by jessegreenberg** [comment]
Above commit is a new implementation for the balloon interaction.  It should be reviewed and discussed with the team, but it works nicely assuming the screen reader supports aria-live, role=application, and role=button.  Rather than switching focus, it uses the solution described here https://github.com/phetsims/balloons-and-static-electricity/issues/148#issuecomment-221402249.  It is a way to support 
> We need to add an interaction alert that first announces the release, and then perspective 3 description. All this before the “Grab Yellow Balloon, button” is read out.

**Comment by terracoda** [comment]
Great @jessegreenberg. Is there a new link?

**Comment by jessegreenberg** [comment]
>  There is currently no indication that the balloon is released.

Checking this off above commits implement a trial solution for this.  It can be reviewed/discussed soon.

**Comment by jessegreenberg** [comment]
> Question: Do we know where the word "undefined" is coming from?

Checking this off as well, new implementation for balloon interraction fixes this problem.  If we see 'undefined' again or need to revert we can look into this.

**Comment by jessegreenberg** [comment]
> The description of the grabbed balloon can be changed to a list consisting of: ...

Checking this off.  Above commits implement this list as a queue of polite alerts.  This works well with our prototype and JAWS but does not work for most other screen readers.  We will need to discuss if this is acceptable, or if we want to come up with a different solution.  NVDA does not seem to support queuing polite alerts at all.

**Comment by jessegreenberg** [comment]
Sent in email, but to address https://github.com/phetsims/balloons-and-static-electricity/issues/148#issuecomment-222025767, here is a link with these changes.

http://www.colorado.edu/physics/phet/dev/html/balloons-and-static-electricity/1.2.0-reader.3/balloons-and-static-electricity_en.html?reader

This is definitely not finished, descriptions are missing and some are still static, but this version addresses everything in the initial issue comment.  Further details will be discussed in a meeting, and new issues will be created for new points of review.

Creating new issue for https://github.com/phetsims/balloons-and-static-electricity/issues/148#issuecomment-221342254.

Closing.
